### PR TITLE
Implement missing parts of embedded objects

### PR DESCRIPTION
**Pull request by tgoyne**
Only Object::create() was implemented, and mutating an object or list after creation didn't work. Migrations involving embedded tables didn't work correctly, and some schema validation was missing.

This changes the SDK API for embedded objects a bit. unbox_embedded() and create_embedded_obectj() are gone, and instead the context must implement create_embedded_object(), which uses the object and property passed to the context at construction time. CreatePolicy is now a struct of bools because there's an additional bit of data that needs to be passed along (whether we're in a create operation or add operation) which determines how we handle nested objects which are already managed by the realm (inside create() we copy existing managed embedded objects, but setting a link directly to a managed embedded object throws). Cocoa previously tracked this state externally, but in a way that didn't work for embedded objects.

This depends on a version of core newer than the latest release, so it won't build successfully on CI.

### CODE REVIEW

File: src/object_schema.cpp [hunk 1]
```diff
@@ -63,27 +63,27 @@ ObjectSchema::ObjectSchema(std::string name, IsEmbedded is_embedded, std::initia
...
+        case col_type_Double:    return PropertyType::Double | flags;
+        case col_type_Bool:      return PropertyType::Bool | flags;
+        case col_type_String:    return PropertyType::String | flags;
+        case col_type_Binary:    return PropertyType::Data | flags;
+        case col_type_Timestamp: return PropertyType::Date | flags;
+        case col_type_OldMixed:  return PropertyType::Any | flags;
+        case col_type_ObjectId:  return PropertyType::ObjectId | flags;
+        case col_type_Decimal:   return PropertyType::Decimal | flags;
+        case col_type_Link:      return PropertyType::Object | PropertyType::Nullable;
+        case col_type_LinkList:  return PropertyType::Object | PropertyType::Array;
```
Comment (ironage) [inline src/object_schema.cpp]: Potentially we could now add a case for converting `col_type_BackLink` into a `PropertyType::LinkingObjects` but that's not really part of this PR (not blocking).

File: src/object_schema.cpp [hunk 2]
```diff
@@ -63,27 +63,27 @@ ObjectSchema::ObjectSchema(std::string name, IsEmbedded is_embedded, std::initia
     }
 }
 
-PropertyType ObjectSchema::from_core_type(Table const& table, ColKey col)
+PropertyType ObjectSchema::from_core_type(ColKey col)
```
Comment (ironage) [inline src/object_schema.cpp]: not having to pass around the table is a nice simplification all around the codebase 💯

File: src/object_schema.cpp (hunk 1 above)
Comment (ironage) [reply src/object_schema.cpp]: We also have `col_type_OldStringEnum` but that should not be possible to reach since AFAIK if we had any enum columns they should have all been converted to string columns in the migration to core 6.

File: src/object.hpp [hunk 3]
```diff
@@ -31,16 +31,41 @@ namespace _impl {
...
+/// Options for how objects should be unboxed by a context.
+///
+/// unbox<Obj>() is used for several different operations which want an Obj
+/// from a SDK type. CreatePolicy packs together all of the different options
+/// around what unbox() should do.
+struct CreatePolicy {
+    /// If given something that is not a managed Object, should an object be
+    /// created in the Realm? False for pure lookup functions such as find(),
+    /// index_of() and queries, true for everything else.
+    bool create : 1;
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply src/object.hpp]: It's not something that the compiler can legally do automatically. Packing them into a bitfield means that you can't take the address of the members, and isn't strictly better performance-wise.

Actually, it's probably not useful to pack them into a bitfield here as we don't have arrays of this type and it fits into a register either way.
